include/libtorrent/aux_/session_impl.hpp

```diff
@@ -118,8 +118,40 @@ namespace libtorrent
 		class item;
 	}
 
-	struct listen_socket_t
+	struct session_udp_socket : utp_socket_interface
 	{
+		session_udp_socket(io_service& ios)
+			: sock(ios) {}
+
+		virtual udp::endpoint local_endpoint() override { return sock.local_endpoint(); }
+
+		udp_socket sock;
+
+		// this is true when the udp socket send() has failed with EAGAIN or
+		// EWOULDBLOCK. i.e. we're currently waiting for the socket to become
+		// writeable again. Once it is, we'll set it to false and notify the utp
+		// socket manager
+		bool write_blocked = false;
+	};
+
+	struct outgoing_udp_socket : session_udp_socket
+	{
+		outgoing_udp_socket(io_service& ios)
+			: session_udp_socket(ios) {}
+
+		// the name of the device the socket is bound to, may be empty
+		// if the socket is not bound to a device
+		std::string device;
+	};
+
+	struct listen_socket_t : dht::dht_socket, aux::session_listen_socket
```

Comment:
The main reason I make a separate type for `dht_socket` was to maintain loose coupling between the dht and session implementations. Now that I look back at it I could just make a separate type not tied to either module.